L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc

```diff
@@ -0,0 +1,266 @@
+                                                  std::array< std::array<std::vector<edm::Ptr<l1t::HGCalTriggerCell>>,40>,2> & reshuffledTriggerCells 
+    ){
+
+    for( edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc = triggerCellsPtrs.begin(); tc != triggerCellsPtrs.end(); ++tc){
+        int endcap = (*tc)->zside() == -1 ? 0 : 1 ;
+        HGCalDetId tcDetId( (*tc)->detId() );
+        int subdet = tcDetId.subdetId();
+        int layer = -1;
+
+        if( subdet == HGCEE ){ 
+            layer = (*tc)->layer();
+        }
+        else if( subdet == HGCHEF ){
+            layer = (*tc)->layer() + 28;
+        }
+        else if( subdet == HGCHEB ){
+            edm::LogWarning("DataNotFound") << "WARNING: the BH trgCells are not yet implemented";            
+        }
+        
+        reshuffledTriggerCells[endcap][layer-1].emplace_back(*tc);
+        
+    }
+
+}
+
+
+/* merge clusters that have common neighbors */
+void HGCalClusteringImpl::mergeClusters( l1t::HGCalCluster & main_cluster, 
+                                         const l1t::HGCalCluster & secondary_cluster ) const
+{
+
+    const edm::PtrVector<l1t::HGCalTriggerCell>& pertinentTC = secondary_cluster.triggercells();
+    
+    for( edm::PtrVector<l1t::HGCalTriggerCell>::iterator tc = pertinentTC.begin(); tc != pertinentTC.end(); ++tc ){
+        main_cluster.addTriggerCell(*tc);
+    }
+
+}
+
+
+void HGCalClusteringImpl::NNKernel( const std::vector<edm::Ptr<l1t::HGCalTriggerCell>> & reshuffledTriggerCells,
+                                    l1t::HGCalClusterBxCollection & clusters,
+                                    const HGCalTriggerGeometryBase & triggerGeometry
+    ){
+   
+    /* declaring the clusters vector */
+    std::vector<l1t::HGCalCluster> clustersTmp;
+
+    // map TC id -> cluster index in clustersTmp
+    std::unordered_map<uint32_t, int> cluNNmap;
+
+    /* loop over the trigger-cells */
+    for( const auto& tc_ptr : reshuffledTriggerCells ){
+                
+        if( tc_ptr->mipPt() < triggerCellThreshold_ ){
+            continue;
+        }
+        
+        // Check if the neighbors of that TC are already included in a cluster
+        // If this is the case, add the TC to the first (arbitrary) neighbor cluster
+        // Otherwise create a new cluster
+        bool createNewC2d(true);
+        const auto neighbors = triggerGeometry.getNeighborsFromTriggerCell(tc_ptr->detId());
+        for( const auto neighbor : neighbors ){
+            auto tc_cluster_itr = cluNNmap.find(neighbor);
+            if(tc_cluster_itr!=cluNNmap.end()){ 
+                createNewC2d = false;
+                try{
+                    clustersTmp.at(tc_cluster_itr->second).addTriggerCell(tc_ptr);
+                    // map TC id to the existing cluster
+                    cluNNmap.emplace(tc_ptr->detId(), tc_cluster_itr->second);
+                }
+                catch(const std::out_of_range& e){
```

Comment:
It's generally inadvisable to have `try/catch` in the normal flow of code. It would be better to check if `tc_cluster_itr->second` is in the range of the vector *before* calling `at()`, and just throw the exception if it's out of range.